feg/gateway/services/n7_n40_proxy/n7/model_conversion.go

```diff
@@ -0,0 +1,567 @@
+	}
+}
+
+func BitRateStringToUint32(br *sbi.BitRateRm) uint32 {
+	if br == nil {
+		return 0
+	}
+	convInt, _ := strconv.ParseUint(string(*br), 10, 32)
+	return uint32(convInt)
+}
+
+func N5qiToQci(n5Qi *sbi.N5Qi) protos.FlowQos_Qci {
+	if n5Qi == nil {
+		return protos.FlowQos_QCI_0
+	}
+	return protos.FlowQos_Qci(*n5Qi)
+}
+
+func GetEventTriggersInfoProto(
+	triggers *[]n7_sbi.PolicyControlRequestTrigger,
+	revalTimeout *time.Time,
+) ([]protos.EventTrigger, *timestamp.Timestamp) {
+	var (
+		protoRevalidationTime *timestamp.Timestamp
+		protoTriggers         []protos.EventTrigger
+	)
+	if triggers == nil {
+		return protoTriggers, protoRevalidationTime
+	}
+	for _, trigger := range *triggers {
+		switch trigger {
+		case n7_sbi.PolicyControlRequestTriggerRETIMEOUT:
+			protoTriggers = append(protoTriggers, protos.EventTrigger_REVALIDATION_TIMEOUT)
+			protoRevalidationTime = ConvertToProtoTimeStamp(revalTimeout)
+		}
+	}
+	return protoTriggers, protoRevalidationTime
+}
+
+func GetUsageMonitorsProto(
+	request *protos.CreateSessionRequest,
+	tgppCtx *protos.TgppContext,
+	eventTriggers []protos.EventTrigger,
+	revalidationTime *timestamp.Timestamp,
+	smPolicyDecision *n7_sbi.SmPolicyDecision,
+) []*protos.UsageMonitoringUpdateResponse {
+	var monitors []*protos.UsageMonitoringUpdateResponse
+	// Usage monitorting at pcc rule level
+	if smPolicyDecision.PccRules != nil {
+		for _, pccRule := range smPolicyDecision.PccRules.AdditionalProperties {
+			if pccRule.RefUmData == nil || len((*pccRule.RefUmData)[0]) == 0 {
+				continue
+			}
+			umEntry, found := smPolicyDecision.UmDecs.Get((*pccRule.RefUmData)[0])
+			if !found {
+				continue
+			}
+			monitors = append(monitors, &protos.UsageMonitoringUpdateResponse{
+				Credit:           GetUsageMonitoringCreditsProto((*pccRule.RefUmData)[0], &umEntry, protos.MonitoringLevel_PCC_RULE_LEVEL),
+				SessionId:        request.SessionId,
+				TgppCtx:          tgppCtx,
+				Sid:              request.GetCommonContext().GetSid().Id,
+				Success:          true,
+				EventTriggers:    eventTriggers,
+				RevalidationTime: revalidationTime,
+			})
+		}
+	}
+	// Usage monitoring at session level
+	if smPolicyDecision.SessRules != nil {
+		for _, sessRule := range smPolicyDecision.SessRules.AdditionalProperties {
+			if sessRule.RefUmData == nil || len(*sessRule.RefUmData) == 0 {
+				continue
+			}
+			umEntry, found := smPolicyDecision.UmDecs.Get(*sessRule.RefUmData)
+			if !found {
+				continue
+			}
+			monitors = append(monitors, &protos.UsageMonitoringUpdateResponse{
+				Credit:           GetUsageMonitoringCreditsProto(*sessRule.RefUmData, &umEntry, protos.MonitoringLevel_SESSION_LEVEL),
+				SessionId:        request.SessionId,
+				TgppCtx:          tgppCtx,
+				Sid:              request.GetCommonContext().GetSid().Id,
+				Success:          true,
+				EventTriggers:    eventTriggers,
+				RevalidationTime: revalidationTime,
+			})
+		}
+	}
+
+	return monitors
+}
```

Comment:
please remember about avoiding code repetition. If you find yourself with repeated code, then you should refactor that. 

It looks like you could create a function to take of extracting monitors out of  `AdditionalProperties`

```
appendToMonitors(AdditionalProperties type(?), monitoringLevel protos.Monitoring_Level, []*protos.UsageMonitoringUpdateResponse)  
```

You can pass the slice and append in the function itself

Reply:
SessRules.AdditionalProperties and PccRules.AdditionalProperties are of different types. But common code can be moved to a new function

Reply:
Hey @GANESH-WAVELABS , per what I see both `AdditionalProperties` are the same object  (I can be wrong though)
```
AdditionalProperties map[string]externalRef1.PresenceInfo 
```

[here](https://github.com/magma/magma/blob/36455a93824e3c93c08d06351c73762c1c4fa02e/feg/gateway/sbi/specs/TS29514NpcfPolicyAuthorization/TS29514NpcfPolicyAuthorization.gen.go#L314:2) 

If you use intellij (free version) you ll be able to follow the definition to the ts file automatically

Reply:
[This](https://github.com/magma/magma/blob/5584180910f1847b56c3e6c410b443640c194f97/feg/gateway/sbi/specs/TS29512NpcfSMPolicyControl/TS29512NpcfSMPolicyControl.gen.go#L813) is type of AdditionalProperties in SessRules and [this](https://github.com/magma/magma/blob/5584180910f1847b56c3e6c410b443640c194f97/feg/gateway/sbi/specs/TS29512NpcfSMPolicyControl/TS29512NpcfSMPolicyControl.gen.go#L793) is the type of AdditionalProperties in PccRules.

In OpenAPI/Swagger AdditionalProperties represents a map/dict like object. The key will be string and the value type can be defined in the openapi file.

Reply:
you are right, they are not the same. I got fooled by the name.
Thanks for adding the links